### ISSUE TITLE
NPE in Clazz.parseClassFile() on analyzing self-referencing annotations

### ISSUE THREAD

**Issue by trespasserw** [bndtools/bnd#1187]
```
java.lang.NullPointerException
  at aQute.bnd.osgi.Clazz.parseClassFile(Clazz.java:747)
  at aQute.bnd.osgi.Clazz.parseClassFile(Clazz.java:512)
  at aQute.bnd.osgi.Clazz.parseClassFileWithCollector(Clazz.java:501)
  at aQute.bnd.component.AnnotationReader.getDef(AnnotationReader.java:131)
  at aQute.bnd.component.AnnotationReader.getDefinition(AnnotationReader.java:127)
  at aQute.bnd.component.DSAnnotations.analyzeJar(DSAnnotations.java:79)
  at aQute.bnd.osgi.Analyzer.doPlugins(Analyzer.java:658)
  at aQute.bnd.osgi.Analyzer.analyze(Analyzer.java:213)
  at aQute.bnd.osgi.Builder.analyze(Builder.java:379)
  at aQute.bnd.osgi.Analyzer.calcManifest(Analyzer.java:688)
```

Happen on processing an annotation class containing a method annotated by the same annotation (e.g. [org.springframework.core.annotation.AliasFor](https://github.com/spring-projects/spring-framework/blob/master/spring-core/src/main/java/org/springframework/core/annotation/AliasFor.java) from spring-core) by DSAnnotations or MetatypeAnnotations plugins.

A test case for ClazzTest (requires DSAnnotations.Options to be public):
```java
	public @interface RecursiveAnno {
		@RecursiveAnno("recursive")
		String value() default "";
	}

	public void testRecursiveAnnotation() throws Exception {
		File file = IO.getFile("bin/test/ClazzTest$RecursiveAnno.class");
		Analyzer analyzer = new Analyzer();
		Clazz clazz = new Clazz(analyzer, file.getPath(), new FileResource(file));
		clazz.parseClassFile();
		analyzer.getClassspace().put(clazz.getClassName(), clazz);
		AnnotationReader.getDefinition(clazz, analyzer, 
			EnumSet.noneOf(DSAnnotations.Options.class),
			new XMLAttributeFinder(analyzer));
	}
```

bndlib 3.0.0.201509101326
Git-Descriptor: 3.0.0.REL
Git-SHA: 0ac3b76c9bd573dfc30cbeaa53700c13ee5187d0

**Comment by trespasserw** [comment]
Caused by recursive invocation of Clazz.parseClassFile(). The inner one resets `pool` to `null` before returning, then the outer one fails on attempt to further process the class.

**Comment by trespasserw** [comment]
The following patch makes Clazz recursion-tolerant, but I'm not sure if it's a correct way.
```patch
Index: biz.aQute.bndlib/src/aQute/bnd/osgi/Clazz.java
===================================================================
--- biz.aQute.bndlib/src/aQute/bnd/osgi/Clazz.java	(revision 664d999b946916bf2c3b2b80417245ae223341ea)
+++ biz.aQute.bndlib/src/aQute/bnd/osgi/Clazz.java	(revision )
@@ -11,10 +11,12 @@
 import java.util.Collection;
 import java.util.Collections;
 import java.util.Comparator;
+import java.util.Deque;
 import java.util.EnumSet;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.LinkedHashMap;
+import java.util.LinkedList;
 import java.util.Map;
 import java.util.Properties;
 import java.util.Set;
@@ -436,6 +438,9 @@
 	boolean	hasClassAnnotations;
 	boolean	hasDefaultConstructor;
 
+	int						depth		= 0;
+	Deque<ClassDataCollector> cds		= new LinkedList<>();
+
 	TypeRef				className;
 	Object				pool[];
 	int					intPool[];
@@ -489,16 +494,20 @@
 	public Set<TypeRef> parseClassFile(InputStream in, ClassDataCollector cd) throws Exception {
 		DataInputStream din = new DataInputStream(in);
 		try {
+			cds.push(this.cd);
 			this.cd = cd;
 			return parseClassFile(din);
 		}
 		finally {
-			cd = null;
+			this.cd = cds.pop();
 			din.close();
 		}
 	}
 
 	Set<TypeRef> parseClassFile(DataInputStream in) throws Exception {
+		analyzer.trace("parseClassFile(): path=%s resource=%s", path, resource);
+
+		++depth;
 		xref = new HashSet<TypeRef>();
 
 		boolean crawl = cd != null; // Crawl the byte code if we have a
@@ -1682,9 +1691,11 @@
 	 */
 
 	public void reset() {
+		if (--depth == 0) {
-		pool = null;
-		intPool = null;
-		xref = null;
+			pool = null;
+			intPool = null;
+			xref = null;
+		}
 	}
 
 	public boolean is(QUERY query, Instruction instr, Analyzer analyzer) throws Exception {
```

**Comment by bjhargrave** [comment]
Perhaps you can make your patch in the form of a pull request (make sure to Signed-off-by your commit comment)? Then it is will be easier for use to review and incorporate your fix.

Please also include the test case changes in the PR.

**Comment by bjhargrave** [comment]
Fixed by #1189
